## Programmatic Assessments in pandas
### Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

### Assess
These are the programmatic assessment methods in pandas that you will probably use most often:

* .head (DataFrame and Series)
* .tail (DataFrame and Series)
* .sample (DataFrame and Series)
* .info (DataFrame only)
* .describe (DataFrame and Series)
* .value_counts (Series only)
* Various methods of indexing and selecting data (.loc and bracket notation with/without boolean indexing, also .iloc)

Try `.head` and `.tail` on the `patients` table.

##### patients table

In [3]:
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

In [4]:
patients.tail(3)

patient_id assigned_sex given_name      surname  \
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [5]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [6]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [7]:
patients.sample(5)

patient_id assigned_sex given_name    surname              address  \
174         175       female   Michelle     Tansey     284 Mudlick Road   
237         238         male       John        Doe      123 Main Street   
450         451         male    Clinton     Miller  901 Southern Street   
502         503         male        Pat    Gersten    2778 North Avenue   
99          100         male      Bernd  Schneider      1262 Deans Lane   

         city     state  zip_code        country  \
174   Spokane        WA   99218.0  United States   
237  New York        NY   12345.0  United States   
450    Roslyn        NY   11576.0  United States   
502      Burr  Nebraska   68324.0  United States   
99   Westbury  New York   11590.0  United States   

                                       contact  birthdate  weight  height  \
174     509-466-2907MichelleDTansey@dayrep.com  2/15/1978   147.8      61   
237                johndoe@email.com1234567890   1/1/1975   180.0      72   
450       516-626-8021ClintonKMiller@rhyta.com  8/16/1985   195.4      74   
502       PatrickGersten@rhyta.com402-848-4923   5/3/1954   138.2      71   
99   BerndSchneider@jourrapide.com914-830-3940  3/24/1993   212.5      71   

      bmi  
174  27.9  
237  24.4  
450  25.1  
502  19.3  
99   29.6

In [8]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Lương          2
Souza          2
Correia        2
Tạ             2
Nilsen         2
Liễu           2
Tucker         2
Parker         2
Cabrera        2
Collins        2
Batukayev      2
Woźniak        2
Ogochukwu      2
Bùi            2
Lâm            2
Johnson        2
Cindrić        2
Gersten        2
Schiavone      2
Silva          2
Lund           2
Grímsdóttir    2
Dratchev       2
Kadyrov        2
Hueber         2
Berg           2
Aranda         2
              ..
Horvat         1
Thạch          1
Karjalainen    1
Bonami         1
Harold         1
Yao            1
Guerra         1
Mustafa        1
Martinsson     1
Tikhonov       1
Casárez        1
Hopkins        1
Morita         1
Rap            1
Nørgaard       1
Lončar         1
Pérez          1
Bakos          1
Schmitt        1
Boyles         1
Vieira         1
Scholz         1
Khoury         1
Terrazas       1
Zaitseva       1
Német          1
Ekechukwu      1
Mai           

In [28]:
patients.query('given_name == "Jake" or surname == "Jakobsen"')

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
432      Houston        TX   77020.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
432  KarenJakobsen@jourrapide.com1 979 203 0438  11/25/1962   185.2      67   

      bmi  
24   24.4  
29   24.4  
432  29.0

In [29]:
patients.query('given_name == "Elizabeth" or surname == "Knudsen"')

patient_id assigned_sex given_name  surname address city state  zip_code  \
230         231       female  Elisabeth  Knudsen     NaN  NaN   NaN       NaN   

    country contact  birthdate  weight  height   bmi  
230     NaN     NaN  9/23/1976   165.9      63  29.4

In [11]:
patients.address.duplicated().sum()

19

In [12]:
patients.address.value_counts()

123 Main Street                  6
2778 North Avenue                2
648 Old Dear Lane                2
2476 Fulton Street               2
1753 Sherman Street              1
87 Wood Duck Drive               1
4870 Corbin Branch Road          1
1324 Duffy Street                1
115 Frank Avenue                 1
1168 Stout Street                1
4038 Farland Street              1
1774 George Avenue               1
4064 Marigold Lane               1
649 Nutter Street                1
4236 Beech Street                1
883 Oakwood Circle               1
204 Rockwell Lane                1
2775 Single Street               1
1116 Dog Hill Lane               1
2074 Parrish Avenue              1
4380 Grim Avenue                 1
4839 North Avenue                1
3686 Meadowcrest Lane            1
3216 Lodgeville Road             1
4019 Cerullo Road                1
2651 Black Oak Hollow Road       1
385 Walnut Avenue                1
4567 Hazelwood Avenue            1
1435 Limer Street   

In [13]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [24]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [25]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [14]:
patients[patients['city'] == 'New York'].shape

(18, 14)

In [15]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

Try `.sample` on the `treatments` table.

##### treatments

In [16]:
treatments.sample(3)

given_name     surname auralin    novodra  hba1c_start  hba1c_end  \
88       yumena    nakayama       -  34u - 32u         7.76       7.35   
57   klementyna  sokołowska       -  42u - 41u         7.98       7.53   
105      finlay    sheppard       -  31u - 30u         7.51       7.17   

     hba1c_change  
88           0.91  
57           0.95  
105          0.34

Try `.info` on the `treatments` table.

In [17]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


Try `.describe` on the `patients` table.

In [18]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [26]:
sum(treatments.auralin.isnull())

0

In [27]:
sum(treatments.novodra.isnull())

0

Try `.value_counts` on the *adverse_reaction* column of the `adverse_reactions` table.

In [21]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [22]:
adverse_reactions.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top         berta  johnson     hypoglycemia
freq            1        2               19

In [23]:
adverse_reactions.adverse_reaction.value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
throat irritation             2
nausea                        2
cough                         2
Name: adverse_reaction, dtype: int64

Try selecting the records in the `patients` table for patients that are from the *city* New York.

#### Quality

##### patients table

* zip code is a float not a strig
* for some registries zip code has four digits
* Tim Neudorf height is 27 instead of 71 inches
* Sometimes there's full state name and other times just the abbreviations
* patient_id 9 has name with typo: Dsvid Gustafson
* missing demographic information (address - contact columns)
* there are 4 columns with wrong data types: assigned_sex, state, zip_code and birthdate)
* different phone number formats
* John Doe is duplicated
* multiple records for Jakobsen, Gersten, Taylor

##### treatements table

* missing HbA1c changes
* the letter u in starting and ending doses for Auralin and Novodra
* lowercase given names and surnames
* missing records (280 instead of 350)
* wrong data types for auralin and novodra columns
* inaccurate value for HbA1c changes (4s mistaken by 9s)
* nulls represented as dashes `-` in auralin and novodra columns

##### adverse_reactions table

* lowercase given names and surnames

#### Tidiness

##### patients table

* `contact` column should be split into phone number and email

##### treatements table

* `auralin` and `novodra` should have two columns, one for minimum and other for maximum dose
* `auralin` and `novodra` are different types of treatment

##### adverse_reactions table

* this table is part of the treatment
 